File: android/app/src/test/java/com/gradethread/app/inventory/SizeCheckTest.kt
package com.gradethread.app.inventory

import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * US-2921: the size-versus-measurements check, run against the SAME two fixture
 * cases as `services/edge-functions/src/tests/size-check_test.ts`,
 * `src/lib/size-check.test.ts` and `ios/GradeThreadTests/SizeCheckTests.swift`.
 * Four copies of one rule only stay one rule if they answer the same questions
 * with the same numbers.
 *
 * THE MOTIVATING CASE. A Lululemon men's top measuring 17.5 in pit to pit,
 * labelled Large. The brand's own chart puts a Large at a 41-43 in body chest,
 * which is a 22-26.5 in flat garment. 17.5 is below the smallest size they make,
 * so the check must fire and say so.
 *
 * THE NO-FALSE-ALARM CASE. An ordinary men's tee measuring 22 in pit to pit,
 * labelled L, on the generic chart. A generic Large is exactly 22-26.5 in flat,
 * so the check must stay quiet. This is the case that keeps the feature usable:
 * a checker that cries wolf on correctly sized items gets switched off, and then
 * it catches nothing at all.
 */
class SizeCheckTest {

    private fun row(size: String, index: Int, chest: List<Double>) =
        SizeCheck.BandRow(size = size, index = index, bands = mapOf("chest" to chest))

    /** What GET /api/flipdesk/size-bands returns for Lululemon men's tops. */
    private val lululemonMensTops = listOf(
        row("XS", 0, listOf(18.0, 22.5)),
        row("S", 1, listOf(19.0, 23.5)),
        row("M", 2, listOf(20.5, 25.0)),
        row("L", 3, listOf(22.0, 26.5)),
        row("XL", 4, listOf(23.5, 28.0)),
        row("XXL", 5, listOf(25.0, 29.5)),
    )

    /** The generic men's alpha fallback, used when a brand has no chart. */
    private val genericMensTops = listOf(
        row("S", 0, listOf(19.0, 23.5)),
        row("M", 1, listOf(20.5, 25.0)),
        row("L", 2, listOf(22.0, 26.5)),
        row("XL", 3, listOf(23.5, 28.0)),
        row("XXL", 4, listOf(25.0, 29.5)),
    )

    // ── The two fixture cases ───────────────────────────────────────────────

    @Test
    fun `motivating case - 17_5 inch chest labelled Large fires`() {
        assertEquals(3, SizeCheck.resolveRow(lululemonMensTops, "Large"))
        val verdict = SizeCheck.check(
            rows = lululemonMensTops,
            rowIndex = 3,
            measurements = mapOf("chest" to 17.5),
            tier = "brand",
        )
        assertEquals(SizeCheck.Status.OFF, verdict.status)
        assertTrue("stepsOff was ${verdict.stepsOff}", verdict.stepsOff >= 2)
        assertEquals("smaller than XS", verdict.impliedSize)
        assertEquals("chest", verdict.key)
        assertEquals(listOf(22.0, 26.5), verdict.expected)
    }

    @Test
    fun `no false alarm - a real 22 inch Large tee stays quiet`() {
        assertEquals(2, SizeCheck.resolveRow(genericMensTops, "L"))
        val verdict = SizeCheck.check(
            rows = genericMensTops,
            rowIndex = 2,
            measurements = mapOf("chest" to 22.0),
            tier = "generic",
        )
        assertEquals(SizeCheck.Status.OK, verdict.status)
        assertEquals(0, verdict.stepsOff)
    }

    // ── Tolerance ───────────────────────────────────────────────────────────

    @Test
    fun `tolerance is one step on a real chart and two on a generic one`() {
        assertEquals(1, SizeCheck.toleranceForTier("verified"))
        assertEquals(1, SizeCheck.toleranceForTier("brand"))
        assertEquals(2, SizeCheck.toleranceForTier("generic"))
    }

    @Test
    fun `a one step disagreement fires on a brand chart and not a generic one`() {
        val onBrand = SizeCheck.check(
            rows = genericMensTops,
            rowIndex = 2,
            measurements = mapOf("chest" to 20.5),
            tier = "brand",
        )
        assertEquals(1, onBrand.stepsOff)
        assertEquals(SizeCheck.Status.OFF, onBrand.status)

        val onGeneric = SizeCheck.check(
            rows = genericMensTops,
            rowIndex = 2,
            measurements = mapOf("chest" to 20.5),
            tier = "generic",
        )
        assertEquals(SizeCheck.Status.OK, onGeneric.status)
    }

    // ── Label matching ──────────────────────────────────────────────────────

    @Test
    fun `size labels resolve across the spellings sellers actually use`() {
        assertEquals(3, SizeCheck.resolveRow(lululemonMensTops, "Large"))
        assertEquals(3, SizeCheck.resolveRow(lululemonMensTops, "l"))
        assertEquals(3, SizeCheck.resolveRow(lululemonMensTops, "  L  "))
        assertEquals(5, SizeCheck.resolveRow(lululemonMensTops, "2XL"))
        assertEquals(5, SizeCheck.resolveRow(lululemonMensTops, "XXL"))
        assertEquals(0, SizeCheck.resolveRow(lululemonMensTops, "extra small"))
    }

    @Test
    fun `no match is null, never row zero`() {
        assertNull(SizeCheck.resolveRow(lululemonMensTops, "42R"))
        assertNull(SizeCheck.resolveRow(lululemonMensTops, ""))
        assertNull(SizeCheck.resolveRow(lululemonMensTops, null))
    }

    @Test
    fun `a bare twelve is not a UK twelve`() {
        val uk = listOf(
            SizeCheck.BandRow("UK 10 / S", 0, mapOf("waist" to listOf(14.0, 16.0))),
            SizeCheck.BandRow("UK 12 / M", 1, mapOf("waist" to listOf(15.0, 17.0))),
        )
        assertEquals(1, SizeCheck.resolveRow(uk, "M"))
        assertEquals(1, SizeCheck.resolveRow(uk, "UK 12"))
        // The corpus warns that assuming a bare 12 is a UK 12 is the costliest
        // mistake on these brands.
        assertNull(SizeCheck.resolveRow(uk, "12"))
    }

    // ── Silence ─────────────────────────────────────────────────────────────

    @Test
    fun `unknown when nothing can be judged`() {
        assertEquals(
            SizeCheck.Status.UNKNOWN,
            SizeCheck.check(genericMensTops, null, mapOf("chest" to 21.0), "brand").status,
        )
        assertEquals(
            SizeCheck.Status.UNKNOWN,
            SizeCheck.check(genericMensTops, 2, emptyMap(), "brand").status,
        )
        assertEquals(
            SizeCheck.Status.UNKNOWN,
            SizeCheck.check(genericMensTops, 2, mapOf("chest" to 21.0), "none").status,
        )
        assertEquals(
            SizeCheck.Status.UNKNOWN,
            SizeCheck.check(emptyList(), 0, mapOf("chest" to 21.0), "brand").status,
        )
    }

    // ── The one-click fix ───────────────────────────────────────────────────

    @Test
    fun `no one click fix for a size the brand does not make`() {
        val verdict = SizeCheck.check(
            rows = lululemonMensTops,
            rowIndex = 3,
            measurements = mapOf("chest" to 17.5),
            tier = "brand",
        )
        assertNull(SizeCheck.fixableSize(verdict))
    }

    @Test
    fun `the fix offers the implied size when the brand makes it`() {
        val verdict = SizeCheck.check(
            rows = lululemonMensTops,
            rowIndex = 5,
            measurements = mapOf("chest" to 22.5),
            tier = "brand",
        )
        assertEquals(SizeCheck.Status.OFF, verdict.status)
        assertEquals(verdict.impliedSize, SizeCheck.fixableSize(verdict))
    }

    // ── Department ──────────────────────────────────────────────────────────

    @Test
    fun `department reads men and women and refuses everything else`() {
        assertEquals("Men", SizeCheck.departmentFromText(listOf("Nike Mens Tee")))
        assertEquals("Women", SizeCheck.departmentFromText(listOf("Womens Blouse")))
        // "women" contains "men"; women must win.
        assertEquals("Women", SizeCheck.departmentFromText(listOf("Lululemon women's top")))
        assertNull(SizeCheck.departmentFromText(listOf("Plain cotton tee")))
        assertNull(SizeCheck.departmentFromText(listOf("Boys size 10 hoodie")))
        assertNull(SizeCheck.departmentFromText(listOf(null)))
    }

    // ── Decoding the endpoint's response ────────────────────────────────────

    @Test
    fun `decodes the band table the edge returns`() {
        val raw = """
            {"tier":"brand","brandLabel":"Lululemon","department":"Men","garment":"Tops",
             "sourceUrl":null,"sizeSystem":"alpha","sizeClass":"standard",
             "measurementBasis":"body",
             "rows":[{"size":"XS","index":0,"bands":{"chest":[18,22.5]}}]}
        """.trimIndent()
        val json = Json { ignoreUnknownKeys = true; isLenient = true }
        val response = json.decodeFromString(SizeCheck.BandsResponse.serializer(), raw)
        assertEquals("brand", response.tier)
        assertEquals("Lululemon", response.brandLabel)
        assertEquals("body", response.measurementBasis)
        assertEquals(listOf(18.0, 22.5), response.rows.first().bands["chest"])
    }
}

File: src/lib/size-check.test.ts
import { describe, it, expect } from "vitest";
import {
  checkSize,
  discrepancyNote,
  fixableSize,
  resolveSizeRow,
  tierNote,
  toleranceFor,
  type SizeBandRow,
} from "@/lib/size-check";

// US-2918: the browser lookup, run against the SAME two fixture cases as
// services/edge-functions/src/tests/size-check_test.ts,
// ios/GradeThreadTests/SizeCheckTests.swift and the Android suite.
// src/test/size-check-fixture-parity.test.ts fails if any of the four drifts.
//
// The band tables below are what GET /api/flipdesk/size-bands returns for these
// two queries — body chest converted to flat as (body + ease) / 2, low edge on
// slim ease and high edge on relaxed.

/** Lululemon men's tops, chest 33-35 / 35-37 / 38-40 / 41-43 / 44-46 / 47-49. */
const LULULEMON_MENS_TOPS: SizeBandRow[] = [
  { size: "XS", index: 0, bands: { chest: [18, 22.5] } },
  { size: "S", index: 1, bands: { chest: [19, 23.5] } },
  { size: "M", index: 2, bands: { chest: [20.5, 25] } },
  { size: "L", index: 3, bands: { chest: [22, 26.5] } },
  { size: "XL", index: 4, bands: { chest: [23.5, 28] } },
  { size: "XXL", index: 5, bands: { chest: [25, 29.5] } },
];

/** Generic men's alpha tops — the fallback when a brand has no chart. */
const GENERIC_MENS_TOPS: SizeBandRow[] = [
  { size: "S", index: 0, bands: { chest: [19, 23.5] } },
  { size: "M", index: 1, bands: { chest: [20.5, 25] } },
  { size: "L", index: 2, bands: { chest: [22, 26.5] } },
  { size: "XL", index: 3, bands: { chest: [23.5, 28] } },
  { size: "XXL", index: 4, bands: { chest: [25, 29.5] } },
];

describe("the motivating case", () => {
  it("a 17.5 in flat chest labelled Large fires", () => {
    const rowIndex = resolveSizeRow(LULULEMON_MENS_TOPS, "Large");
    expect(rowIndex).toBe(3);
    const verdict = checkSize({
      bands: LULULEMON_MENS_TOPS,
      rowIndex,
      measurements: { chest: 17.5 },
      tier: "brand",
    });
    expect(verdict.status).toBe("off");
    expect(verdict.stepsOff).toBeGreaterThanOrEqual(2);
    expect(verdict.impliedSize).toBe("smaller than XS");
    expect(verdict.key).toBe("chest");
    expect(verdict.expected).toEqual([22, 26.5]);
  });

  it("names both numbers in the note", () => {
    const verdict = checkSize({
      bands: LULULEMON_MENS_TOPS,
      rowIndex: 3,
      measurements: { chest: 17.5 },
      tier: "brand",
    });
    expect(discrepancyNote(verdict, "Large")).toBe(
      "Measurements point to smaller than XS, not Large. " +
        "A Large usually measures 22 to 26.5 in here.",
    );
  });

  it("offers no one-click fix for a size the brand does not make", () => {
    const verdict = checkSize({
      bands: LULULEMON_MENS_TOPS,
      rowIndex: 3,
      measurements: { chest: 17.5 },
      tier: "brand",
    });
    expect(fixableSize(verdict)).toBeNull();
  });
});

describe("the no-false-alarm case", () => {
  it("a real 22 in men's tee labelled L stays quiet", () => {
    const rowIndex = resolveSizeRow(GENERIC_MENS_TOPS, "L");
    expect(rowIndex).toBe(2);
    const verdict = checkSize({
      bands: GENERIC_MENS_TOPS,
      rowIndex,
      measurements: { chest: 22 },
      tier: "generic",
    });
    expect(verdict.status).toBe("ok");
    expect(verdict.stepsOff).toBe(0);
  });
});

describe("tolerance", () => {
  it("is one step on a real chart and two on a generic one", () => {
    expect(toleranceFor("verified")).toBe(1);
    expect(toleranceFor("brand")).toBe(1);
    expect(toleranceFor("generic")).toBe(2);
  });

  it("a one-step disagreement fires on a brand chart and not on a generic one", () => {
    const input = {
      bands: GENERIC_MENS_TOPS,
      rowIndex: 2,
      measurements: { chest: 20.5 },
    };
    expect(checkSize({ ...input, tier: "brand" }).stepsOff).toBe(1);
    expect(checkSize({ ...input, tier: "brand" }).status).toBe("off");
    expect(checkSize({ ...input, tier: "generic" }).status).toBe("ok");
  });
});

describe("label matching", () => {
  it("resolves the spellings sellers actually use", () => {
    expect(resolveSizeRow(LULULEMON_MENS_TOPS, "Large")).toBe(3);
    expect(resolveSizeRow(LULULEMON_MENS_TOPS, "l")).toBe(3);
    expect(resolveSizeRow(LULULEMON_MENS_TOPS, "  L  ")).toBe(3);
    expect(resolveSizeRow(LULULEMON_MENS_TOPS, "2XL")).toBe(5);
    expect(resolveSizeRow(LULULEMON_MENS_TOPS, "extra small")).toBe(0);
  });

  it("returns null on no match, never index 0", () => {
    expect(resolveSizeRow(LULULEMON_MENS_TOPS, "42R")).toBeNull();
    expect(resolveSizeRow(LULULEMON_MENS_TOPS, "")).toBeNull();
    expect(resolveSizeRow(LULULEMON_MENS_TOPS, null)).toBeNull();
  });

  it("a bare 12 is not a UK 12", () => {
    const uk: SizeBandRow[] = [
      { size: "UK 10 / S", index: 0, bands: { waist: [14, 16] } },
      { size: "UK 12 / M", index: 1, bands: { waist: [15, 17] } },
    ];
    expect(resolveSizeRow(uk, "M")).toBe(1);
    expect(resolveSizeRow(uk, "UK 12")).toBe(1);
    expect(resolveSizeRow(uk, "12")).toBeNull();
  });
});

describe("unknown", () => {
  it("says nothing when the size cannot be placed or nothing compares", () => {
    const cases = [
      { rowIndex: null, measurements: { chest: 21 }, tier: "brand" as const },
      { rowIndex: 2, measurements: {}, tier: "brand" as const },
      { rowIndex: 2, measurements: { chest: 21 }, tier: "none" as const },
    ];
    for (const c of cases) {
      expect(checkSize({ bands: GENERIC_MENS_TOPS, ...c }).status).toBe("unknown");
    }
    expect(
      checkSize({ bands: [], rowIndex: 0, measurements: { chest: 21 }, tier: "brand" }).status,
    ).toBe("unknown");
  });
});

describe("copy", () => {
  it("a generic chart says out loud that it is an estimate", () => {
    expect(tierNote("generic", "Lululemon")).toBe(
      "Estimate only — no Lululemon chart on file.",
    );
    expect(tierNote("generic", null)).toBe("Estimate only — no brand chart on file.");
    expect(tierNote("brand", "Lululemon")).toBeNull();
    expect(tierNote("verified", "Lululemon")).toBeNull();
  });

  it("offers the implied size as a one-click fix when the brand makes it", () => {
    const verdict = checkSize({
      bands: LULULEMON_MENS_TOPS,
      rowIndex: 5,
      measurements: { chest: 22.5 },
      tier: "brand",
    });
    expect(verdict.status).toBe("off");
    expect(fixableSize(verdict)).toBe(verdict.impliedSize);
  });
});
